core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -2582,7 +2586,7 @@ private[spark] class CallerContext(
         val callerContext = Class.forName("org.apache.hadoop.ipc.CallerContext")
         val builder = Class.forName("org.apache.hadoop.ipc.CallerContext$Builder")
         // scalastyle:on classforname
-        val builderInst = builder.getConstructor(classOf[String]).newInstance(context)
+        val builderInst = builder.getConstructor(classOf[String]).newInstance(prepareContext(context))
```

Comment:
The reason prepareContext is called in each setCallerContext is that currently every client only need to set the caller context once, and each CallerContext object only call setCallerContext once. Yes, I agree with your point, that will make the implementation better and benefit future invocations.